icefall/context_graph.py

```diff
@@ -0,0 +1,222 @@
+    in the ContextGraph, we will give the decoded sequence a bonus to make it survive
+    beam search.
+    """
+
+    def __init__(self, context_score: float):
+        """Initialize a ContextGraph with the given ``context_score``.
+
+        A root node will be created (**NOTE:** the token of root is hardcoded to -1).
+
+        Args:
+          context_score:
+            The bonus score for each token(note: NOT for each word/phrase, it means longer
+            word/phrase will have larger bonus score, they have to be matched though).
+        """
+        self.context_score = context_score
+        self.root = ContextState(token=-1, score=0, total_score=0, is_end=False)
+        self.root.fail = self.root
+
+    def _fill_fail(self):
+        """This function fills the fail arc for each trie node, it can be computed
+        in linear time by performing a breadth-first search starting from the root.
+        See https://en.wikipedia.org/wiki/Aho%E2%80%93Corasick_algorithm for the
+        details of the algorithm.
+        """
+        queue = []
+        for token, node in self.root.next.items():
+            node.fail = self.root
+            queue.append(node)
+        while queue:
+            current_node = queue.pop(0)
+            current_fail = current_node.fail
+            for token, node in current_node.next.items():
+                fail = current_fail
+                if token in current_fail.next:
+                    fail = current_fail.next[token]
+                node.fail = fail
+                queue.append(node)
+
+    def build_context_graph(self, token_ids: List[List[int]]):
+        """Build the ContextGraph from a list of token list.
+        It first build a trie from the given token lists, then fill the fail arc
+        for each trie node.
+
+        See https://en.wikipedia.org/wiki/Trie for how to build a trie.
+
+        Args:
+          token_ids:
+            The given token lists to build the ContextGraph, it is a list of token list,
+            each token list contains the token ids for a word/phrase. The token id
+            could be an id of a char (modeling with single Chinese char) or an id
+            of a BPE (modeling with BPEs).
+        """
+        for tokens in token_ids:
+            node = self.root
+            for i, token in enumerate(tokens):
+                if token not in node.next:
+                    node.next[token] = ContextState(
+                        token=token,
+                        score=self.context_score,
+                        # The total score is the accumulated score from root to current node,
+                        # it will be used to calculate the score of fail arc later.
+                        total_score=node.total_score + self.context_score,
+                        is_end=i == len(tokens) - 1,
+                    )
+                node = node.next[token]
+            self._fill_fail()
```

Comment:
@pkufool I think the building of failure arcs should be global, not per-phrase basis. The idea is to scan the entire list, figuring out the "proper suffix" relations between all trie nodes. If the failure arcs are built per-phrase basis, in the middle of the graph building, existing nodes can not see nodes from unseen(not-built-yet) phrases.

Reply:
Oh, yes, the indent is wrong, it is very strange that such a error did not affect the performance. Thanks!